### PR TITLE
Add royalty fee support

### LINKED ISSUE #1918: FCOTMR views can become inconsistent on a `CryptoTransfer` rollback.
### Description

Suppose a `CryptoTransfer` leads to two `BalanceChange`s where, in order:
1. The first changes ownership of an NFT.
2. The second tries to transfer units of a FT to an account not associated to the FT.

The second `BalanceChange` will trigger a call to `HederaLedger.dropPendingTokenChanges()`; so that the pending changes in the `nftsLedger` should not committed. **However**, the FCOTMR changes made in [`HederaTokenStore.changeOwner()`](https://github.com/hashgraph/hedera-services/blob/master/hedera-node/src/main/java/com/hedera/services/store/tokens/HederaTokenStore.java#L309) will not be rolled back!

### Steps to reproduce

_(Will be added shortly.)_

### Version

v0.17.0-alpha

### LINKED ISSUE #1919: NFTs ledger not rolled back in `HederaLedger.dropPendingTokenChanges()`
### Description

The [`HederaLedger.dropPendingTokenChanges()` method](https://github.com/hashgraph/hedera-services/blob/master/hedera-node/src/main/java/com/hedera/services/ledger/HederaLedger.java#L363) should also call `nftsLedger.rollback()`.

### Version

v0.17.0-alpha

### LINKED ISSUE #1935: Add `RoyaltyFee` to custom HTS fee types
### Problem

NFT creators cannot define a royalty fee to be automatically assessed when their NFTs are exchanged for fungible value (whether ℏ or units of a fungible HTS token like USDC).

Neither the `FixedFee` nor `FractionalFee` types support this use case.

### Solution

Introduce a new `RoyaltyFee` type:
```
/* A fee to assess during a CryptoTransfer that changes ownership of an NFT. 
*
* Defines the fraction of the fungible value exchanged for an NFT that the ledger 
* should collect as a royalty. 
*   - "Fungible value" includes both ℏ and units of fungible HTS tokens.
*   - If the NFT sender does not receive any fungible value, the ledger will assess the (optional) fallback fee to the new NFT owner. 
*/  

message RoyaltyFee {
  Fraction exchange_value_fraction = 1; // The fraction of fungible value exchanged for an NFT to collect as royalty
  FixedFee fallback_fee = 2; // If present, the fixed fee to assess to the NFT receiver when no fungible value is exchanged with the sender
}
```

Implement the charging behavior outlined in the two examples [here](https://github.com/hashgraph/hedera-services/blob/update-custom-fee-examples/docs/fees/custom-fees-characterization.md#royalty-fees) where first,
> Zephyr sends Amelie an NFT of type westWindArt (with no other transfers), where the westWindsArt token has a royalty fee of 1/100th exchange value and a fallback fee of 1USDC.

And then,
> Amelie sends Alice the westWindArt NFT, but this time in exchange for 1000ℏ and 200USDC.

### LINKED ISSUE #1972: Use `_WITH_CUSTOM_FEES` price types not only when `feeScheduleKey` is set
### Description

At present the `_WITH_CUSTOM_FEES` price type is only triggered when the `feeScheduleKey` is set.

### Steps to reproduce

Create a token with custom fees, but without a `feeScheduleKey`.

### Additional context

_No response_

### Hedera network

other

### Version

v0.17.0-SNAPSHOT

### Operating system

_No response_

### PR DESCRIPTION
**Pull request by tinker-michaelj**
**Description**:
- Implement the `RoyaltyFee` type described in #1935.
- Correctly rollback all NFT-related changes from [both `accountsLedger` and `tokenViewsManager`](https://github.com/hashgraph/hedera-services/blob/01935-M-AddRoyaltyFee/hedera-node/src/main/java/com/hedera/services/ledger/HederaLedger.java#L379) in `HederaLedger.dropPendingTokenChanges()`.
- Switch to `_WITH_CUSTOM_FEES` subtypes for a `TokenCreate` w/ custom fees but no `feeSchedulesKey`.

**Related issue(s)**:
- Closes #1935
- Closes #1918 
- Closes #1919
- Closes #1972

**Notes for reviewer**:
New cases studies are described in the custom fee charging doc.

### CODE REVIEW

File: hedera-node/src/main/java/com/hedera/services/grpc/marshalling/BalanceChangeManager.java [hunk 1]
```diff
@@ -37,6 +39,7 @@
 	private int levelNo = 0;
 	private int levelStart = 0;
 	private int levelEnd;
+	private Set<Pair<Id, Id>> royaltiesPaid = null;
```
Comment (qnswirlds) [inline hedera-node/src/main/java/com/hedera/services/grpc/marshalling/BalanceChangeManager.java:42]: should initialize this `royaltiesPaid` to `new HashSet<>()` and remove the `if` on lines 56-58.

Comment (tinker-michaelj) [reply hedera-node/src/main/java/com/hedera/services/grpc/marshalling/BalanceChangeManager.java:42]: Creating this unconditionally would affect *every* `CryptoTransfer`, even xfers that will never touch NFTs.

**Review by ljianghedera — APPROVED** [review]
Overall looks awesome!

**Review by qnswirlds — DISMISSED** [review]
LGTM

File: test-clients/src/main/java/com/hedera/services/bdd/suites/token/TokenUpdateSpecs.java [hunk 2]
```diff
@@ -96,30 +98,30 @@ public static void main(String... args) {
...
+//						invalidTreasuryCheckHolds(),
+//						newTreasuryMustSign(),
+//						newTreasuryMustBeAssociated(),
+//						tokensCanBeMadeImmutableWithEmptyKeyList(),
+//						updateHappyPath(),
+//						updateNftTreasuryHappyPath(),
+//						updateTokenTreasuryRequiresZeroTokenBalance(),
+//						validatesMissingAdminKey(),
+//						validatesMissingRef(),
+//						validatesNewExpiry(),
```
Comment (abhishek-hedera) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/token/TokenUpdateSpecs.java:121]: Should these tests be commented out?

Comment (tinker-michaelj) [reply test-clients/src/main/java/com/hedera/services/bdd/suites/token/TokenUpdateSpecs.java]: Nope 😝 , fixed!

**Pull request merged**